src/ess/amor/utils.py

```diff
@@ -70,49 +74,226 @@ def theta_grid(
+        (None,) * n
+        if arg is None
+        else (arg,) * n
+        if isinstance(arg, sc.Variable)
+        else arg
+    )
+
+
+def wavelength_theta_figure(
+    da: sc.DataArray | Sequence[sc.DataArray],
+    *,
+    wavelength_bins: (sc.Variable | None) | Sequence[sc.Variable | None] = None,
+    theta_bins: (sc.Variable | None) | Sequence[sc.Variable | None] = None,
+    q_edges_to_display: Sequence[sc.Variable] = (),
+    linewidth: float = 1.0,
+    **kwargs,
+):
+    '''
+    Creates a figure displaying a histogram over :math:`\\theta` and :math:`\\lambda`.
+
+    The input can either be a single data array containing the data to display, or
+    a sequence of data arrays.
+
+    The inputs must either have coordinates called "theta" and "wavelength",
+    or they must be histograms with dimensions "theta" and "wavelength".
+
+    If :code:`wavelength_bins` or :code:`theta_bins` are provided, they are used
+    to construct the histogram. If not provided, the function uses the
+    bin edges that already exist on the data arrays.
+
+    If :code:`q_edges_to_display` is provided, lines will be drawn in the figure
+    corresponding to :math:`Q` equal to the values in :code:`q_edges_to_display`.
+
+    Parameters
+    ----------
+    da : array or sequence of arrays
+        Data arrays to display.
+    wavelength_bins : array-like, optional
+        Bins used to histogram the data in wavelength.
+    theta_bins : array-like, optional
+        Bins used to histogram the data in theta.
+    q_edges_to_display : sequence of float, optional
+        Values of :math:`Q` to be displayed as straight lines in the figure.
+    linewidth : float, optional
+        Thickness of the displayed :math:`Q` lines.
+    **kwargs : keyword arguments, optional
+        Additional parameters passed to the histogram plot function,
+        used to customize colors, etc.
+
+    Returns
+    -------
+        A Plopp figure displaying the histogram.
+    '''
+
+    if isinstance(da, sc.DataArray):
+        return wavelength_theta_figure(
+            (da,),
+            wavelength_bins=(wavelength_bins,),
+            theta_bins=(theta_bins,),
+            q_edges_to_display=q_edges_to_display,
+            **kwargs,
+        )
+
+    wavelength_bins, theta_bins = (
+        _repeat_variable_argument(len(da), arg) for arg in (wavelength_bins, theta_bins)
+    )
+
+    hs = []
+    for d, wavelength_bin, theta_bin in zip(
+        da, wavelength_bins, theta_bins, strict=True
+    ):
+        bins = {}
+        if wavelength_bin is not None:
+            bins['wavelength'] = wavelength_bin
+
+        if theta_bin is not None:
+            bins['theta'] = theta_bin
+        elif (
+            'theta' not in d.dims
+            and 'sample_rotation' in d.coords
+            and 'detector_rotation' in d.coords
+        ):
+            bins['theta'] = theta_grid(
+                nu=d.coords['detector_rotation'], mu=d.coords['sample_rotation']
+            )
```

Comment:
This code is also repeated below. Factor it out also?

Basically, it feels like these 3 functions are doing something pretty similar. They all make a set of images on the same axes. The dimensions of the axes are Q, theta, and/or z. I think we could reduce duplication quite a bit more?
Also the parts with `setdefault` and `imagefigure`?

Reply:
I extracted the part of the code that tries to create a theta grid if it wasn't provided.
I don't think it makes sense to de-duplicate the rest.